Backpack command

Sfiguz7 opened the pull request:
## Description
Added a command that allows for backpack retrieval in the event they are lost. The command accepts a name and id, if those match up it spawns a Medium Backpack with the correct lore set in the sender's inventory - if not it sends them a message about which of the two isn't found.

## Changes
Added a new SubCommand class, edited messages-en.yml to include 2 new error messages and and the command description, edited plugin.yml to add the command

## Related Issues
N/A

## Checklist
- [ V ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ = ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
Only ExtraTools and TranscEndence were loaded with it, it does not mess with anything Addon related though.
- [ X ] I followed the existing code standards and didn't mess up the formatting.
Under Walshy's pressure (so blame him) I rewrote the SubCommand to have if/return checks instead of the nested ifs other SubCommands follow.
- [ X ] I did my best to add documentation to any public classes or methods I added.
Not sure what "proper" documentation would imply for this class so I assume not.

WalshyDev commented:
First off, checkboxes can be ticked with `[X]`.

Secondly,
> Under Walshy's pressure
If it was under my pressure you'd also have all those variables as final. 

You also should really give the correct backpack. You can't give Soulbound but that's fine as the admin can get a Soulbound rune anyway. However, they can't just change it from a medium to large or small.

Sfiguz7 commented:
Boi you could've said that when I told you I was ditching the switch :(
Checkboxes I ticked differently to differentiate between yes no and kindof, as I explained in a one liner just below. I'll happily make the variables final too for you <3

TheBusyBiscuit commented:
Alternatively (since the rest of the itemmeta doesn't matter) you could give them a "Restored Backpack" or "Generic Backpack" item which simply serves as a reference to a lost backpack, can't upgrade that one though of course.

Also going by size could lead to issues since coolers or dyed backpacks also classify as backpacks but share the same size as other backpacks.

WalshyDev commented:
really should just be an id field in the backpack section of data

TheBusyBiscuit commented on src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BackpackCommand.java:
```diff
@@ -0,0 +1,76 @@
+        Player p = (Player) sender;
+        String ownerName = args[1];
+        int id = Integer.parseInt(args[2]);
+        OfflinePlayer owner = Bukkit.getPlayer(ownerName);
+
+        if (owner == null) {
+            owner = Bukkit.getOfflinePlayer(ownerName);
```
This is a little redundant. Bukkit.getOfflinePlayer(...) will return an online player if possible anyway, so potentially calling this twice makes little sense, can pretty much just call Bukkit.getOfflinePlayer(...) right away

TheBusyBiscuit commented on src/main/resources/plugin.yml:
```diff
@@ -37,6 +37,9 @@ permissions:
   slimefun.command.versions:
     description: Allows you to do /sf versions
     default: op
+  slimefun.command.backpack:
+    description: Allows you to do /sf backpack
+    default: op
```
Also make sure to add the new permission and command to the wiki

TheBusyBiscuit commented on src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/Commands.java:
```diff
@@ -25,5 +25,6 @@ public static void addCommands(SlimefunCommand cmd, Collection<SubCommand> comma
         commands.add(new OpenGuideCommand(plugin, cmd));
         commands.add(new SearchCommand(plugin, cmd));
         commands.add(new DebugFishCommand(plugin, cmd));
+        commands.add(new BackpackCommand(plugin,cmd));
```
Please add a space after the comma

TheBusyBiscuit commented on src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BackpackCommand.java:
```diff
@@ -0,0 +1,76 @@
+            PlayerBackpack bp = profile.getBackpack(id).get();
+            //No point switching for size, would still miss soulbound
+            ItemStack item = SlimefunItems.BACKPACK_MEDIUM;
+            ItemMeta meta = item.getItemMeta();
+            List lore = meta.getLore();
+            lore.set(2, ChatColor.GRAY + "ID: " + profile.getUUID().toString() + "#" + id);
+            meta.setLore(lore);
+            item.setItemMeta(meta);
```
Pretty sure you can just use BackpackListener#setBackpackId(...) or so here, might have to look into that

TheBusyBiscuit commented on src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BackpackCommand.java:
```diff
@@ -0,0 +1,76 @@
+            lore.set(2, ChatColor.GRAY + "ID: " + profile.getUUID().toString() + "#" + id);
+            meta.setLore(lore);
+            item.setItemMeta(meta);
+            Slimefun.runSync(() -> p.getInventory().addItem(item));
```
Should probably also send a message like "Heres your backpack" or so

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BackpackCommand.java:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        final Player p = (Player) sender;
+        final String ownerName = args[1];
```
Variable is only ever used once, no point in it

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BackpackCommand.java:
```diff
@@ -0,0 +1,66 @@
+            return;
+        }
+        if (args.length != 3) {
+            SlimefunPlugin.getLocal().sendMessage(sender, "messages.usage", true, msg -> msg.replace("%usage%", "/sf backpack <Player> <ID>"));
```
Looks very long, I assume this is over column limit?

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BackpackCommand.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public void onExecute(CommandSender sender, String[] args) {
+        if (!(sender instanceof Player) || !(sender.hasPermission("slimefun.command.backpack"))) {
```
Redundant brackets around hasPermission statement

WalshyDev commented on src/main/java/me/mrCookieSlime/Slimefun/Lists/SlimefunItems.java:
```diff
@@ -72,6 +72,8 @@ private SlimefunItems() {}
     public static final ItemStack RADIANT_BACKPACK = new SlimefunItemStack("RADIANT_BACKPACK", "40cb1e67b512ab2d4bf3d7ace0eaaf61c32cd4681ddc3987ceb326706a33fa", "&eRadiant Backpack", "", "&7Size: &e54 (Double chest)", "&7ID: <ID>", "", "&7&eRight Click&7 to open");
     public static final SlimefunItemStack BOUND_BACKPACK = new SlimefunItemStack("BOUND_BACKPACK", "2a3b34862b9afb63cf8d5779966d3fba70af82b04e83f3eaf6449aeba", "&cSoulbound Backpack", "", "&7Size: &e36", "&7ID: <ID>", "", "&7&eRight Click&7 to open");
     public static final SlimefunItemStack COOLER = new SlimefunItemStack("COOLER", "d4c1572584eb5de229de9f5a4f779d0aacbaffd33bcb33eb4536a6a2bc6a1", "&bCooler", "&rAllows you to store Juices/Smoothies", "&rand automatically consumes them when you are hungry", "&rand you have this in your Inventory", "", "&7Size: &e27", "&7ID: <ID>", "", "&7&eRight Click&7 to open");
+    public static final ItemStack RESTORED_BACKPACK = new SlimefunItemStack("RESTORED_BACKPACK", "40cb1e67b512ab2d4bf3d7ace0eaaf61c32cd4681ddc3987ceb326706a33fa", "&eRestored Backpack", "", "&7Retrieve your lost items", "", "&7&eRight Click&7 to open");
```
Make this a SlimefunItemStack

Sfiguz7 replied:
Why? All of them are ItemStack, it would not be coherent

WalshyDev replied:
They're being moved to SlimefunItemStack. ItemStack is the old shit. You're right now just causing stupidly useless casts.

WalshyDev commented on src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BackpackCommand.java:
```diff
@@ -0,0 +1,67 @@
+                return;
+            }
+            Slimefun.runSync(() -> {
+                p.getInventory().addItem(SlimefunItems.RESTORED_BACKPACK);
```
Never set the backpack ID or anything...

WalshyDev commented:
Be careful formatting SF... you did did over 6000 line changes

Sfiguz7 commented:
How the...? What did I do/can it be prevented?

WalshyDev commented:
Just dont press your formatting key on a class you know for a fact isn't formatted well. 

You'll need to revert the last commit on that file.

TheBusyBiscuit commented on src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BackpackCommand.java:
```diff
@@ -0,0 +1,77 @@
+            }
+            Slimefun.runSync(() -> {
+                ItemStack item = SlimefunItems.RESTORED_BACKPACK;
+                ItemMeta meta = item.getItemMeta();
+                List lore = meta.getLore();
+                lore.set(2, ChatColor.GRAY + "ID: " + profile.getUUID().toString() + "#" + id);
+                meta.setLore(lore);
+                item.setItemMeta(meta);
```
Pretty sure you can do BackpackListener#setBackpackId(...) her... wait a second didn't I already say that?

Sfiguz7 replied:
I think I mistakenly resolved it and didn't read it heh now it's resolved

TheBusyBiscuit commented on src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/tools/HiddenBackpack.java:
```diff
@@ -0,0 +1,14 @@
+import me.mrCookieSlime.Slimefun.api.SlimefunItemStack;
+import org.bukkit.inventory.ItemStack;
+
+public class HiddenBackpack extends SlimefunBackpack {
```
Should maybe rather be called "RestoredBackpack". Also please add adequate javadocs here.

TheBusyBiscuit commented on src/main/resources/languages/messages_en.yml:
```diff
@@ -42,6 +43,12 @@ guide:
 
   cheat:
     no-multiblocks: '&4You cannot cheat in Multiblocks, you have to build them!'
+
+  backpack:
+    invalid-id: '&4The id must be a non-negative integer!'
+    player-never-joined: '&4No player with that name has ever joined the server!'
+    backpack-does-not-exist: '&4That backpack does not exist!'
+    restored-backpack-given: 'Backpack restored successfully! Added to your inventory!'
```
Why exactly is this in the "guide" section? Doesn't seem to have anything to do with the Slimefun guide to me.
Should probably rather be in the "commands" section.

TheBusyBiscuit commented on src/main/resources/languages/messages_en.yml:
```diff
@@ -42,6 +43,12 @@ guide:
     no-multiblocks: '&4You cannot cheat in Multiblocks, you have to build them!'
+
+  backpack:
+    invalid-id: '&4The id must be a non-negative integer!'
```
Rather use the word "number", "integer" does not appear to be so user-friendly. Also maybe call it "backpack id" rather than just "id"

TheBusyBiscuit commented on src/main/resources/languages/messages_en.yml:
```diff
@@ -42,6 +43,12 @@ guide:
+    invalid-id: '&4The id must be a non-negative integer!'
+    player-never-joined: '&4No player with that name has ever joined the server!'
+    backpack-does-not-exist: '&4That backpack does not exist!'
+    restored-backpack-given: 'Backpack restored successfully! Added to your inventory!'
```
Maybe some color? :eyes:

Sfiguz7 replied:
RED! 😈

TheBusyBiscuit commented on src/main/resources/languages/messages_en.yml:
```diff
@@ -15,6 +15,12 @@ commands:
     reset: '&cYou have reset %player%''s Knowledge'
     reset-target: '&cYour Knowledge has been reset'
+
+  backpack:
```
This is actually a duplicate key. You already placed the description under "commands.backpack".

TheBusyBiscuit commented on src/main/java/io/github/thebusybiscuit/slimefun4/core/commands/subcommands/BackpackCommand.java:
```diff
@@ -0,0 +1,72 @@
+                return;
+            }
+            Slimefun.runSync(() -> {
+                ItemStack item = SlimefunItems.RESTORED_BACKPACK;
```
Hold up...
Very dangerous here, you need to add a .clone() otherwise you are modifying the original and not a copy.

TheBusyBiscuit commented on src/main/resources/languages/messages_en.yml:
```diff
@@ -17,6 +16,7 @@ commands:
     reset-target: '&cYour Knowledge has been reset'
 
   backpack:
+    description: Retrieve an existing backpack
```
You updated the key though but you don't seme to use it actually :?

Sfiguz7 replied:
Doesn't /sf help retrieve that field(?)

TheBusyBiscuit replied:
How would it know what field to retrieve though...?
You just changed the standard location for it so you need to override getDescription() too.

Sfiguz7 replied:
Hmm idk I tried /sf help and it showed the correct one so I assumed it worked. I'll review it then

Sfiguz7 replied:
I see it now, overridden getDescription(); I'll wipe my folder just in case and test, then push

Sfiguz7 commented:
When I changed it to BackpackListener#setBackpackID I passed the wrong player; that method requires a Player though and owner is only an OfflinePlayer so IntelliJ gets mad.
Should I revert to having the 5 lines explicitly? Would you prefer if I changed the required parameter to an OfflinePlayer instead?
